### DISCUSSION TITLE
Thoughts and questions about building a public and open licence course

### DISCUSSION THREAD

**Discussion by samuelroland** [heig-vd-dai-course/heig-vd-dai-course#180] [General]
Hello @ludelafo and @hadrylouis,

(I hope it's okay to discuss this here as I'm in the other DAI class, but I had a few thoughts and questions about this course as an external observer).

As a curious student who want to bring improvements in IT teaching at the HEIG-VD, I dug into the various content types you provide in this repository and I find excellent to have just everything in public and under a free culture licence: the slides, the training material, the archived tests, the practical works, diagram source files, quiz files, ... It's so frustrating with a lot of HEIG-VD courses to wait for exos solutions to come, to not being able to access tests corrections, ...

I read some of the discussions in the Discussions panel too and I find your teaching approach very interesting. We are a lot to be bored by long lecture based courses, so I like the idea of having only a short theory part then to practice and experiment during most of the time. And you seem to want to continuously improve with regular feedback from students, I heard that this is appreciated.

:smiley: As a student I definitely see the advantages as I could contribute, improve or clarify some sections, propose new exercises, build and share exercises upon the existing questions and explanations... It's frustrating when a labo is very confusing since years and no one could change it. Mostly because students don't have access to source files and an easy way to submit them propositions. Not needing to log in on Moodle, download files and put them in the right folder is a big win too. Having slides in text files is a very interesting approach, even it requires a bit of upfront effort I guess (configure the CI/CD and have a nice export).

I would love to see other teachers picking the same approach, especially on choosing to manage their course with Git, in a public repos and under a CC-BY-SA licence. Before talking to some teachers about it, I was wondering a few things:
1. What were your motivations to build the course in a public git repository ? 
1. What were the additional benefits on the teachers' side ? In terms of course edition or deployment ?
1. What were your considerations on the free licence itself ? (Maybe between CC-BY-SA and CC-BY ?)
1. Do teachers own the copyright on the content they create or is it the property of the school ? In the second case, have you been through a special process to ask for permissions to the direction before applying the licence ?
1. Were you expecting/pushing contributions from your students ?
1. Why didn't you choose the traditional approach of building slides with PowerPoint and exporting PDF ? Would you recommend Marp to other teachers or is there any better text based slides tool you discovered in the meantime ?

:question: I know it's a lot of questions, but your answers could really help me convince other teachers to do the same, so thank you in advance !

Samuel Roland

PS: if there are other students wanting to participate and share their opinions, you are welcome !

**Comment by ludelafo** [comment]
Hello @samuelroland,

Thank you very much for all your comments.

It is a large topic and I could talk about this for hours. Before I continue, I must stress this:

- What I'm about to share with you concerns only my own experience and my own opinion of what I think is a good teaching experience according to my own core values.
- I have no training in teaching. I do think I have a facility for it, but I do not have training in it. A lot of things are done instinctively and according to the feedback I get from students. What I know comes mainly from the help of the school's educational consultant and my own experience.
- I do not think my course/my approach is perfect. Far from this. I know some mistakes were made and I see many things to improve for next year, technically and pedagogically (#171).
- I did not work alone. @hadrylouis was an **invaluable** help for the preparation of this course. He helped me by reading-proof all the course material, giving feedback and inputs when elements were not clear and adding tremendous value to the course. Other people helped as well.
- I know this approach is not easy for everyone in the class: I am a perfectionist and I know that I am demanding of my students.
- I sometimes fear that this approach was not a good idea: I took a risk that could have been very costly if it did not work well. I did it by challenge because that is the kind of course I would have wanted and I only hope that this will pay on the long run, for myself but above all for my students.

What you see with this course is the result of _many_ years of frustration as an ex-student and as a software engineer in my professional life.

My main philosophy with computer sciences (and many other fields) is this:

- **Straight to the point**: I do not want to waste anyone's time. It has to be quick, efficient and useful.
- **Open**: People can improve/change/ask for help about the content I provide. 
- **Resilient**: People should be able to take over whatever I did with the less friction as possible. Too many times I have seen projects thrown away and "started over" because no one could understand what has been done.
- **Freedom**: People should be able to do what they like to do. Doing something personal, in a fun and interesting way, is the best way to engage and to learn something new. I do not have a right to judge other people's idea.
- **Shit happens**: People can do mistakes. People can have ups and downs. People can have opinions. It is life. I want to set a safe space where people can come to me to discuss at any time if they want. If I can listen/help/improve, I will.

This all this in mind, let's start with your questions.

> What were your motivations to build the course in a public git repository ?

I wanted to show/express how to use tools that are/can be used in the industry, in the real life. Git is vastly used in the industry, and I wanted to reflect this in my course: you will learn something useful for yourself, for your (future) jobs, for working with others, etc.

These tools/methodologies must reflect what is used in the real life, not something that is forgotten after the end of the course.

> What were the additional benefits on the teachers' side ? In terms of course edition or deployment ?

The main thing I like is that everything is in one place: the course material, the code, the discussions/questions, etc.

Back in my days, we had resources on the NAS share, some in CyberLearn, some on GitHub, announcements by mails, etc. It was a big mess, a mental burden. I wanted something easy to use, something easy to understand with everything in one place. I try to use mail/Teams only for one-to-one discussions with my students if it cannot be done face-to-face.

I must admit I am not a big fan of GitHub but it is the standard right now. Everyone uses it and we should learn it as soon as possible. It offers everything I need for my course and if my students can learn to use it correctly, I think it will be useful in their future.

> What were your considerations on the free licence itself ? (Maybe between CC-BY-SA and CC-BY ?)

I like to think that education should be easily available to anyone. Other big universities (EPFL, MIT, and many others) offer their courses for free and I wanted to do the same: not only my students could learn from this now but in the future as well (as this course is meant to evolve over time) and I think it shows that the HEIG-VD is open to the rest of the world. I think it is a good promotion for our school, the course and the quality we can produce in our _beautiful_ walls.

Using the CC-BY-SA license, I try to protect this status quo (from others or from myself) for future usage.

> Do teachers own the copyright on the content they create or is it the property of the school ? In the second case, have you been through a special process to ask for permissions to the direction before applying the licence ?

I do not have all the elements on the legal aspect of this point of but I did not ask for permission.

> Were you expecting/pushing contributions from your students ?

I do not expect contributions from my students. If there are contributions, I am more than happy to merge them with the current course but I do not expect them.

I wanted to give people the chance to have a voice in its structure/creation/maintenance. I also wanted to give people the chance to make quick and easy suggestions for improvements on the course.

If something is wrong/not clear, anyone could ask for changes, either by opening an Issue or a Discussion.

> Why didn't you choose the traditional approach of building slides with PowerPoint and exporting PDF ? Would you recommend Marp to other teachers or is there any better text based slides tool you discovered in the meantime ?

First things first, I must state this: I have learned from the school's educational consultant that a PowerPoint presentation is not a course material ("_une présentation n'est pas un support de cours_").

Before becoming aware of this, I had this dilemma:

- _Do I make slides that are nice to see but I do need to talk a lot to explain the content?_
- _Do I make slides that have many things on them so everything is available for the students to learn on their own?_

In the first case, I do not see myself talking for 1h30 straight to give my course, and I remember that I really did not like it when I was a student: I was bored too quickly to continue to listen. In the second case, in the end, if I only read the slides, it is useless; people can read it on their own, they do not need me to read all the slides one after the other for them. Issues I had as a student myself.

Another question I had was _how to structure the practical content I want the students to do_. In a presentation, it quickly becomes a mess and I did not see how to overcome this only using some slides.

Becoming aware of the "_une présentation n'est pas un support de cours_", I ditched the idea of only using presentations/slides for my course.

This is where the idea of the presentations + course material came alive: I can quickly present the big picture through slides that try to illustrate the topic I want to talk about and students then switch to a course material where they get their hands dirty. People can get deeper in the course on their own and I am then available to help students that might have further questions/need more explanations. It is a win-win situation in my book.

The main idea behind this is that I wanted to allow people to go at their own pace: if someone wants to start with the theoretical part of the course before starting the practical content, they can. If someone wants to "learn by doing" (like me), they can. A mix between the two is also possible. It also enables the hybrid on-site/off-site work: if people cannot come to my course, it is not a problem, they have everything they need to learn on their own, at home, at another time, etc. I really try to avoid the "but I said it in class, you were not here/were not listening, that's why you did not get the info". Everything must be written somewhere so everyone can have access to it.

Back to the aspect of PowerPoint.

PowerPoint (and "binary" files in general) are super hard to work with and to version/keep track of changes in my opinion. There are many things I considered when I planned to make presentations:

- I wanted something that is visually appealing. However, I do not consider myself good for this so I needed something I could customize for my needs once and for all and then forget about it. I still do not know if my slides are easy to grasp but at least they are not over-filled with images/diagrams/texts/concepts/etc. (in my opinion).
- I did not want to lose time arranging elements in my slides: the content is more important than the form and I do not want to waste time for this. I needed a tool to quickly build and present concepts. I only needed a few features: a title, some content, images on one side of the slides if needed, a few bullet points and that's it.
- I needed something that anyone could reuse without learning a specific program. Markdown is all that is needed.
- I needed a program that would still allow me to generate PDFs so students could annotate slides if needed.
- I needed people to be able to access the raw content of the slides/course material so they can translate all resources: I know not everyone is fluent in English. English being the standard in our field, I wanted this course to be an opportunity to improve students English skills and get them into the habit of reading technical documentation that is written in English.
- I wanted to be able to track changes with versioning and being able to make a course easily updated.
- I wanted an open source tool that I could use with a CI/CD pipeline to generate the slides.

I did search for a quite some time for a nice tool to fill all this and quickly came across Marp.

A good friend of mine made me discover [Slidev](https://sli.dev/). Much more powerful but not as straightforward as Marp. It ressembles much more to a Vue project and I did not want to manage another website for my course.

[Reveal.js](https://revealjs.com/) is another alternative many people I know use but I really do not like it for two reasons: I do not want to write HTML for my slides and PDF generation is very limited/not well supported.

In the end, Marp is super super easy to use (a Visual Studio extension to install, a few CSS classes to add to each file to improve the design and you are ready to start) and you can write entire slides using only Markdown and that's it. Even if it is not the best, people could even read the slides on GitHub as it is plain Markdown if they want to.

Another GitHub user has made quite a complete list of Markdown presentation tools [here](https://gist.github.com/johnloy/27dd124ad40e210e91c70dd1c24ac8c8) if needed.

A quick note about the course material PDF generation.

As mentioned, I wanted students to be able to have a PDF copy of the course so they could annotate it if needed. At the moment, the course material PDFs are okayish and I am not really happy with the results. I did try many things to transform the Markdown files to PDF, without great success. The current attempt is to use [Pandoc](https://pandoc.org/) to transform the Markdown to HTML and pass the result to [Weasyprint](https://weasyprint.org/) with the help of [one of their official template](https://github.com/CourtBouillon/weasyprint-samples/tree/master/report) but the result is still not very good. I might need to try the Markdown to LaTeX transformation at some point and see the results.

Hopefully, the reading experience on GitHub is very good in my opinion and should cover the 90% of the needs and is OK for the needs of this course but for other specific courses, I would consider writing the course material with LaTeX or [Typst](https://typst.app).

---

With all that said, I only hope that this approach can inspire other people to try the same. I really do want to see a change in how teaching is done today. I think learning and teaching can be fun. On both sides. Having a more modern approach (at least I have the impression mine is) can only be beneficial. In the end, I only hope my students will enjoy the course and that everyone can get something meaningful out of it.

What I can emphases is this: **if you are not happy about your teaching experience**, if you have any improvements you can share for the courses you follow, **say it**: go talk to the teacher(s) in charge, fill the GAPS evaluations with your opinion, ask for changes, demand the quality you deserve.

I encourage you to continue your work to achieve these changes.

If you need anything or want to talk in person, feel free to hit me up.

**Comment by Nortalle** [comment]
Hello @samuelroland, @ludelafo, @hadrylouis !

What an amazing topic ! 

I teach the WEB course at the HEIG, and I am really interested in the best way to do it.

Just so you know, the people that teach at the HEIG are not teacher to begin with. But we have a lot of resource to become one ! All the training courses proposed by https://www.hes-so.ch/devpro. I am maybe mistaken, but people that teach at the HES now needs to follow 15 courses in less than 2 years to have the credit to continue teaching in HES, along-side other conditions, like 128 periods of teaching and 200 hours of work.

We had a lot of discussions with @ludelafo and two of the core concepts for the moment for what teaching should be like in HES are:

- "[Competence is] a complex knowledge-action based on the effective mobilization and combination of a variety of internal and external resources within a family of situations." - Jacques Tardif.
- "[La compétence est] un savoir-agir complexe prenant appui sur la mobilisation et la combinaison efficaces d'une variété de ressources internes et externes à l'intérieur d'une famille de situations." - Jacques Tardif.
- "Intelligence is not what we know, but what we do when we don't know."
- "L'intelligence n'est pas ce que nous savons, mais ce que nous faisons lorsque nous ne savons pas."

Based on these concepts, it's up to us to create a course that enables people to acquire these skills and, above all, to be able to test them, equally for everyone. And it's best if we add fun and kindness on it.

I cannot wait to read more in this discussion and if there is a talk in person, I'd like to take part !

Vincent Guidoux

**Comment by samuelroland** [comment]
Hello,

Sorry for taking so much time to reply, I was too busy for this end of semester with projects to release and exams, and I wanted to think about it before replying...

First, thank you very much @ludelafo for such a long and detailed answer, I read it several times, thanks for sharing all of these interesting thoughts !

It's very interesting to read the details of your teaching philosophy, that's definitely a question I will ask more often to other teachers in the future. I definitely see the link between the free culture licence and your points *Open* and *Freedom*. I totally understand the main advantages for teachers you explained: using modern tools, having everything in the same place, general education openness and promotion of the school. Thanks for your answers for all my questions.

I agree with the problem of binary files like PowerPoint, and found interesting to know what was the design process around the slides' dilemma. I will try Marp and Slidev soon.

In the meantime, I was very happy to find the GAPS evaluation in the repository because sometimes we don't even see it, or the teacher just scrolls during 5 minutes, but we can't access it after. I read the whole GAPS evaluation and FramaForms CSVs and even if some issues were raised regarding evaluations and work load, it's clear that the course has been globally well received.  
By the way during the DAI exam, I preferred the style of your questions because they required a bit of thinking around real life situations (and not just remembering facts or copying them back from the personal notes sheet). Finally, I hope the frustration that leads you to not follow the status quo and the risk you took by taking this new approach was worth it. As the course is over now, I wonder if you consider it a "success" ?

I'm thinking about suggesting the CC-BY-SA licence to the PRG2 course (for the [new exercises' repository](https://github.com/PRG2-HEIGVD/PRG2_Recueil_Exercices)) but I'm going to wait a few weeks to continue these discussions and have a better understanding on the legal side related to copyright owners.

I'm motivated to talk about this in person, I'll reply to your email very soon !

PS: thanks @Nortalle for contributing to the conversation and the precisions regarding what teachers at the HEIG are required to do. I heard about Jacques Tardiff in the [Pédagoscope podcast](https://pedagoscope.ch/), I guess you know it too ? See you on Wednesday for the WEB course :)

CC: @hadrylouis

**Comment by ludelafo** [comment]
Hello Samuel,

The pleasure is mine.

Yes, I do consider this course a success.

I'm proud of my students (I think all of us learned something new to take for the future), I'm glad the way we (teaching staff and students) worked together as team using the feedback to improve the course, and I'm more than pleased with the current state of the GitHub Organization: if someone (myself included) has to take over or take inspiration of what we did, they can. Every aspect of this work is documented, the thought process is explained (I have even mentioned our discussion for future teachers here: https://github.com/heig-vd-dai-course/heig-vd-dai-course/tree/main/99-material-for-the-teaching-staff) and tools are accessible to work on/improve what was done.

I am looking forward to teaching next year and offer an even-more polished experience. I am also looking forward to seeing the "impact" this way of teaching can bring to the HEIG-VD.

See you in a few hours! :)
